feat: add the sign feature

kayagokalp opened the pull request:
## About this PR
closes #7.

This PR introduces a new command `sign` to the `forc-wallet`. Currently `sign` accepts 2 parameters:

1. Transaction id
2. Account index of the account that we want to sign this transaction with

In this first iteration of development (where we are trying to enable our users to sign deployment transactions) The usual signing process while deploying a contract currently looks like the following:

1. Running `forc-deploy` will ask the user to input the address of the wallet they are going to be signing with. To get that the user can either run the `forc-wallet list` and look under the `address` section of the desired account or run `forc-wallet account <account_index>` to get the address. This reports a transaction id and waits for the signature.
2. User runs `forc-wallet sign <transaction_id> <account_index>` and `forc-wallet` asks for the password for the wallet as we do need to decrypt it before signing. If the password is correct, the signature is generated and reported back to the user.
3. In the meantime `forc-deploy` is waiting for the signature to be inserted, once the user inserts the generated signature from the `forc-wallet` it adds it to the witnesses and submits the transaction.

All the required functionality from the wallet's end is ready with this PR and once https://github.com/FuelLabs/sway/pull/2629 is merged `forc-deploy` will also behave as described.

JoshuaBatty reviewed (commented):
Nice one. Looks reasonable, will defer to @mitchmindtree

kayagokalp commented:
I am converting this to draft, because just now I managed to sign a transaction and while doing so I had to make some changes. I will update the description while pushing them.

kayagokalp commented:
This is ready to go! I added the plain address section to the `list` command but I believe the naming there could be improved. Since in the current implementation `forc-deploy` does not communicate with `forc-wallet` as a lib, we do need a way of providing the wallet that we are going to be signing the transaction as `forc-deploy` will prepare the deployment transaction and provide the id to the user (1st step in the description). So the address that we are seeing under the `address` section of the `forc-wallet list` (or `forc-wallet account <account_index>`) is for that. 

The newly added `plain address` section is very helpful while testing as to sign a contract we need to add coins to that account through chain config. In the "chain config" we need to provide the plain address, so I added that section to the `list` command.

I can also improve the CLI experience (with some suggestions 😅)  in this PR too, but if you prefer reaching a point where we can complete the whole signing process first and iterate over it for better UX I am also down for that too.

adlerjohn commented on src/account.rs:
```diff
@@ -43,6 +42,7 @@ pub(crate) fn new_account(path: Option<String>) -> Result<()> {
     create_accounts_file(&wallet_path, account_addresses)?;
 
-    println!("Wallet public address: {}", wallet.address());
+    println!("Wallet address: {}", wallet.address());
+    println!("Wallet plain address: {}", wallet.address().hash());
```
Is the plain address the hash of the pubkey? I don't think that should be printed out by default. Regular end-users should be using bech32 addresses exclusively. Pubkey hash should be something users need to purposefully seek out as advanced users.

kayagokalp replied:
I added hash of the pubkey, since it is useful/needed while testing in local as (if i am not mistaken) chain config requires that for funding the wallet with initial coins. But this is indeed something an advanced user would use. Maybe we can introduce a flag to `account` command in a future PR such that executing something like `forc-wallet account <account_index> --hashed` would return this.

adlerjohn commented on src/list.rs:
```diff
@@ -19,10 +23,15 @@ pub(crate) fn print_wallet_list(path: Option<String>) -> Result<(), Error> {
+            "[{}] {} 0x{}",
+            index,
+            address,
+            Bech32Address::from_str(&address)?.hash()
```
ditto

adlerjohn commented on src/sign.rs:
```diff
@@ -0,0 +1,23 @@
+    };
+    let tx_id = Bytes32::from_str(id).map_err(|e| anyhow!("{}", e))?;
+    let secret_key = derive_account_with_index(&wallet_path, account_index)?;
+    let message = unsafe { Message::from_bytes_unchecked(*tx_id) };
```
For clarity, this should be called message hash?

mitchmindtree commented on src/utils.rs:
```diff
@@ -62,3 +63,14 @@ pub(crate) fn number_of_derived_accounts(path: &Path) -> usize {
+    )?;
+    let phrase_recovered = eth_keystore::decrypt_key(path.join(".wallet"), password)?;
+    let phrase = String::from_utf8(phrase_recovered)?;
+    let derive_path = format!("m/44'/1179993420'/{}'/0/0", account_index);
```
Perhaps we should be using an upstream `const` and/or function for constructing this path? E.g. at least a `wallet_account_derive_path` function or similar. It's probably fine for now, but I've opened an issue to follow-up: https://github.com/FuelLabs/forc-wallet/issues/28.

mitchmindtree reviewed (approved):
Just tested this locally with [your `forc deploy` branch](https://github.com/FuelLabs/sway/pull/2629) and it appears to be working nicely!
